snuba_sdk/utils.py

```diff
@@ -0,0 +1,138 @@
+        else:
+            agg = to_exp(a)
+            selected_columns.append(agg)
+
+    arrayjoin = body.get("arrayjoin")
+    if arrayjoin:
+        selected_columns.append(Function("arrayJoin", [Column(arrayjoin)], arrayjoin))
+
+    query = query.set_select(selected_columns)
+
+    groupby = body.get("groupby", [])
+    if groupby and not isinstance(groupby, list):
+        groupby = [groupby]
+
+    query = query.set_groupby(list(map(to_exp, groupby)))
+
+    conditions = []
+    for cond in body.get("conditions", []):
+        if len(cond) != 3 or not isinstance(cond[1], str):
+            raise InvalidQuery("OR conditions not supported yet")
+
+        conditions.append(Condition(to_exp(cond[0]), Op(cond[1]), to_scalar(cond[2])))
+
+    extra_conditions = ["project", "organization"]
+    for cond in extra_conditions:
+        column = Column(f"{cond}_id")
+        values = body.get(cond)
+        if isinstance(values, int):
+            conditions.append(Condition(column, Op.EQ, values))
+        elif isinstance(values, list):
+            rhs: Sequence[Any] = list(map(to_scalar, values))
+            conditions.append(Condition(column, Op.IN, rhs))
+        elif isinstance(values, tuple):
+            rhs = tuple(map(to_scalar, values))
+            conditions.append(Condition(column, Op.IN, rhs))
+
+    date_conds = [("from_date", Op.GT), ("to_date", Op.LTE)]
+    for cond, op in date_conds:
+        date_str = body.get(cond, "")
+        if date_str:
+            # HACK: This is to get sessions working quickly.
+            # The time column should depend on the entity.
+            conditions.append(
+                Condition(Column("started"), op, parse_datetime(date_str))
+            )
+
+    query = query.set_where(conditions)
+
+    having = []
+    for cond in body.get("having", []):
+        if len(cond) != 3 or not isinstance(cond[1], str):
+            raise InvalidQuery("OR conditions not supported yet")
+
+        having.append(Condition(to_exp(cond[0]), Op(cond[1]), to_scalar(cond[2])))
+
+    query = query.set_having(having)
+
+    order_by = body.get("orderby")
+    if order_by:
+        if not isinstance(order_by, list):
+            order_by = [order_by]
+
+        order_bys = []
+        for o in order_by:
+            direction = Direction.DESC if o.startswith("-") else Direction.ASC
+            order_bys.append(OrderBy(Column(o.lstrip("-")), direction))
+
+        query = query.set_orderby(order_bys)
+
+    limitby = body.get("limitby")
+    if limitby:
+        limit, name = limitby
+        query = query.set_limitby(LimitBy(Column(name), int(limit)))
+
+    extras = ("limit", "offset", "granularity", "totals")
+    for extra in extras:
+        if body.get(extra) is not None:
+            query = getattr(query, f"set_{extra}")(body.get(extra))
+
+    # sample = legacy.get("sample")
+    # sample_clause = f"SAMPLE {sample}" if sample else ""
```

Comment:
Why are these commented ?

Reply:
I had a section here for things that we don't support yet. I'll remove for now.

Reply:
I did intentionally leave out the SAMPLE clause in order to avoid feature creep. The point of this was to add the bare minimum to support the features used by the sessions queries. Since those don't use sample, I didn't add it.